src/kaleidoscope/MatrixAddr.h

```diff
@@ -0,0 +1,385 @@
+// indices. For most keyboardsw that feature less than 256 indices keys/LEDs/...
+// the row/col information can be stored in a more memory efficient fashion.
+// In that case if saving memory is paramount, use
+// prefer template class MatrixAddrCompressed that is defined later on
+// in this file.
+//
+template<uint8_t rows__, uint8_t cols__>
+class MatrixAddr {
+ private:
+  uint8_t row_;
+  uint8_t col_;
+
+ public:
+
+  static constexpr uint8_t rows = rows__;
+  static constexpr uint8_t cols = cols__;
+
+  static constexpr uint8_t rowSize = cols__;
+
+  static constexpr uint8_t matrixSize = rows__ * cols__;
+
+  static constexpr uint8_t invalidIndex = 255;
+
+  constexpr MatrixAddr() : row_(invalidIndex), col_(invalidIndex) {}
+
+  explicit constexpr MatrixAddr(uint8_t row, uint8_t col) : row_(row), col_(col) {}
+
+  explicit constexpr MatrixAddr(uint8_t offset)
+    : MatrixAddr(offset / cols, offset % cols) {}
+
+  template<typename MatrixAddr__>
+  constexpr MatrixAddr(const MatrixAddr__ &other) : MatrixAddr(other.row(), other.col()) {}
+
+  constexpr uint8_t row() const {
+    return row_;
+  }
+  constexpr uint8_t col() const {
+    return col_;
+  }
+
+  void setRow(uint8_t r) {
+    row_ = r;
+  }
+  void setCol(uint8_t c) {
+    col_ = c;
+  }
+
+  constexpr uint8_t offset() {
+    return row_ * cols + col_;
+  }
+
+  constexpr bool isValid() const {
+    return (row_ < rows__) && (col_ < cols__);
+  }
+
+  template<typename MatrixAddr__>
+  MatrixAddr& operator=(const MatrixAddr__ &other) {
+    this->row_ = other.row();
+    this->col_ = other.col();
+    return *this;
+  }
+
+  template<typename MatrixAddr__>
+  MatrixAddr &operator+(const MatrixAddr__ &other) {
+    this->row_ += other.row();
+    this->col_ += other.col();
+    return *this;
+  }
+
+  template<typename MatrixAddr__>
+  MatrixAddr &operator-(const MatrixAddr__ &other) {
+    this->row_ -= other.row();
+    this->col_ -= other.col();
+    return *this;
+  }
+
+  template<typename MatrixAddr__>
+  MatrixAddr &operator+=(const MatrixAddr__ &other) {
+    *this = *this + other;
+    return *this;
+  }
+
+  template<typename MatrixAddr__>
+  MatrixAddr &operator-=(const MatrixAddr__ &other) {
+    *this = *this - other;
+    return *this;
+  }
+};
+
+// This is a copressed storage version of MatrixAddr.
+// It uses a common integer to store row and column indices.
+// To enable fast row/col access, bitshifts are used instead of
+// integer divisions to compute row indices. This is only possible
+// by padding rows with unused entries to gain a row size that
+// is a power of two.
+//
+template<uint8_t rows__, uint8_t cols__, typename StorageType__ = uint8_t>
+class MatrixAddrCompressed {
+ private:
+
+  StorageType__ raw_;
+
+ public:
+
+  typedef MatrixAddrCompressed<rows__, cols__, StorageType__> ThisType;
+
+  // Row size must be a power of two to enable fast index extraction.
+  static constexpr uint8_t rowSize
```

Comment:
Again, the [coding style guide specifies lower case for variable names](https://github.com/keyboardio/Kaleidoscope/blob/master/doc/CODING_STYLE.md#variable-names). I'm finding it much more difficult than necessary reading this code where variables are using the same style as function names.

Reply:
My experience with many other C++ projects is that a mixture of camel and snake case for symbol names as demanded by the Kaleidoscope style guide is rarely used.

Anyway, i adapted all variable and parameter names to be conforming with the style guide.

Reply:
BTW, I don't like snake case because it breaks the ability to rely on case sensitivity when searching for symbol names. 

For search strings that start with a capital letters, with camel case you can always be sure that you will only get hits at the beginning of a word, e.g. searching for `Bus(` would find `readBus(...)` or similar things, while searching for `bus` in the presence of snake case variables also gives you all hits that end with the syllable bus like `find_succubus(`. That's no problem in an environment where advanced search features like regex are available but it makes a huge difference when browsing code in other places where no such features are available, like e.g. web browsers (github).